nix/redis-cluster.nix

```diff
@@ -0,0 +1,145 @@
+{ pkgs, lib, name, config, ... }:
+let
+  inherit (lib) types;
+in
+{
+  options = {
+    enable = lib.mkEnableOption name;
+
+    package = lib.mkPackageOption pkgs "redis" { };
+
+    dataDir = lib.mkOption {
+      type = types.str;
+      default = "./data/${name}";
+      description = "The redis-cluster data directory (common for all nodes).";
+    };
+
+    port = lib.mkOption {
+      type = types.port;
+      default = 30001;
+      description = ''
+        The TCP port to accept connections.
+
+        Next node will use the next port and so on.
+
+        If port is set to `0`, redis will not listen on a TCP socket.
+      '';
+    };
+
+    bind = lib.mkOption {
+      type = types.nullOr types.str;
+      default = "127.0.0.1";
+      description = ''
+        The IP interface to bind to.
+        `null` means "all interfaces".
+
+        All the nodes will follow the same bind address.
+      '';
+      example = "127.0.0.1";
+    };
+
+    timeout = lib.mkOption {
+      type = types.int;
+      default = 2000;
+      description = ''
+        Time (in milliseconds) after which the node is considered to be down.
+
+        If master node is down, a replica node will take over.
+      '';
+    };
+
+    nodes = lib.mkOption {
+      type = types.int;
+      default = 6;
+      description = ''
+        Number of nodes in the cluster.
+      '';
+    };
+
+    replicas = lib.mkOption {
+      type = types.int;
+      default = 1;
+      description = ''
+        Number of replicas per Master node.
+      '';
+    };
+
+    extraConfigs = lib.mkOption {
+      type = types.attrsOf types.lines;
+      default = { };
+      description = "Attrset map of node (identified by port number) to the respective `redis.conf`.";
```

Comment:
This seems rather unconventional - to have port number as attrset keys? Besides, if someone changes `port` down the line, the whole attrset would get invalidated, no? I may be missing something, but why not something like:

```nix
{
  redis-cluster.nodes = {
    node1 = { port = 30001; extraConfig = "..."; };
    node2 = { port = 30002; extraConfig = "..."; };
  };
}
```

.. which seems rather straightforward to read.

Reply:
This makes more sense. I was trying to replicate the create cluster script provided by redis that allocates nodes incrementally on ports one after another, that's why I ended up doing this as init. Will change the behaviour as per the API you mentioned, which will be cleaner.